### PR TITLE
[DMS-89] OpenSearch query integration

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: eng/deployment/postgresql-kafka-opensearch/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+Tools
+
+* [Kafdrop](http://localhost:9000/)
+* [OpenSearch Dashboard](http://localhost:5601/)
+  * After a record has been inserted, create an [index
+    pattern](http://localhost:5601/app/management/opensearch-dashboards/indexPatterns): `ed-fi$*`
+
+Known problems:
+
+* DELETE is not having any impact on OpenSearch.
```
Comment (stephenfuqua) [inline eng/deployment/postgresql-kafka-opensearch/README.md:44]: If we don't already have a ticket for that, please create one. I suspect that we're not emitting the tombstone event properly from our transform.

File: eng/deployment/postgresql-kafka-opensearch/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# PostgreSQL -> Kafka -> OpenSearch
+
```
Comment (stephenfuqua) [inline eng/deployment/postgresql-kafka-opensearch/README.md:2]: Please add a note briefly describing why this exists (presumably for setting up a functional development environment).

File: src/backend/EdFi.DataManagementService.Backend.OpenSearch/QueryOpenSearch.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                // DeepClone() so they can be placed in a new JsonArray
+                .Select(node => node!["_source"]!["edfidoc"]!.DeepClone())!
+                .ToArray()!;
+
+            return new QueryResult.QuerySuccess(new JsonArray(documents), totalCount);
+        }
+        catch (Exception ex)
+        {
+            logger.LogCritical(ex, "Uncaught Query failure - {TraceId}", queryRequest.TraceId);
+            return new QueryResult.UnknownFailure("Unknown Failure");
```
Comment (stephenfuqua) [inline src/backend/EdFi.DataManagementService.Backend.OpenSearch/QueryOpenSearch.cs:124]: Can you confirm if the API client will get an error response that includes the trace id in it?

File: docs/DOCKER.md [hunk 4]
```diff
@@ -37,6 +37,7 @@ OAUTH_TOKEN_ENDPOINT=<Authentication service url>
 BYPASS_STRING_COERCION=<Boolean whether to bypass coercion of boolean and numeric values represented as strings to their natural type. Eg. "true" = true>
 DATABASE_ISOLATION_LEVEL=<The System.Data.IsolationLevel to use for transaction locking. Eg. RepeatableRead>
 DATABASE_CONNECTION_STRING=<The non-admin database connection string>
+OPENSEARCH_ENDPOINT=<The URL endpoint to the OpenSearch server>
```
[the remark this answers is not in the document]
Comment (bradbanister) [reply docs/DOCKER.md:40]: I'll create a ticket for it. We've got security disabled on OpenSearch, would need to figure out and test out.

File: eng/deployment/postgresql-kafka-opensearch/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    ```
+
+1. Restart the PostgreSQL container.
+1. Create the `edfi_datamanagementservice` database and run the DB installer. 
+1. Run the following SQL commands against the `edfi_datamanagementservice` database:
+    ```sql
+    CREATE PUBLICATION to_debezium WITH (publish = 'insert, update, delete, truncate', publish_via_partition_root = true);
+    SELECT pg_create_logical_replication_slot('debezium', 'pgoutput');
+    ALTER PUBLICATION to_debezium ADD TABLE dms.document;
+    ```
```
Comment (stephenfuqua) [inline eng/deployment/postgresql-kafka-opensearch/README.md:32]: I never took this step. Is this necessary because of the partitions?

Comment (bradbanister) [reply eng/deployment/postgresql-kafka-opensearch/README.md:32]: Yes, this is that fix. `publish_via_partition_root = true`

File: eng/deployment/postgresql-kafka-opensearch/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      - zookeeper
+    environment:
+      - ZOOKEEPER_CONNECT=dms-zookeeper1:2181
+      - KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://dms-kafka1:9092
+    volumes:
+      - kafka-data:/kafka/data
+      - kafka-logs:/kafka/logs
+
+  # Kafka UI on port 8080 - https://docs.kafka-ui.provectus.io/
+  kafka-ui:
```
Comment (bradbanister) [inline eng/deployment/postgresql-kafka-opensearch/docker-compose.yml:32]: Switched from Kafdrop to Kafka UI. Better browsing, and other features.

File: src/appsettings.template.json [hunk 7]
```diff
@@ -1,13 +1,15 @@
 {
     "AppSettings": {
         "AuthenticationService": "${OAUTH_TOKEN_ENDPOINT}",
-        "DatabaseEngine": "postgresql",
+        "Datastore": "postgresql",
+        "QueryHandler": "postgresql",
```
Comment (bradbanister) [inline src/appsettings.template.json:5]: `opensearch` to switch

File: src/backend/EdFi.DataManagementService.Backend.Installer/Program.cs [hunk 8]
```diff
@@ -18,7 +18,7 @@ public class Options
         Required = true,
         HelpText = "The database engine to deploy. Valid values are 'postgresql' and 'mssql'"
     )]
-    public string? DatabaseEngine { get; set; }
+    public string? Datastore { get; set; }
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Installer/Program.cs:21]: Renamed to fit "datastore" vs "query handler" language

File: src/core/EdFi.DataManagementService.Core/Handler/Utility.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+using EdFi.DataManagementService.Core.External.Model;
+
+namespace EdFi.DataManagementService.Core.Handler;
+
+public static class Utility
+{
+    /// <summary>
+    /// Formats a error result string from the given error information and traceId
+    /// </summary>
+    public static string ToJsonError(string errorInfo, TraceId traceId)
```
Comment (bradbanister) [inline src/core/EdFi.DataManagementService.Core/Handler/Utility.cs:16]: Was a string extension, no longer made sense once traceId needed to be there too.

File: eng/deployment/postgresql-kafka-opensearch/README.md (hunk 5 above)
Comment (stephenfuqua) [reply eng/deployment/postgresql-kafka-opensearch/README.md:32]: @msilesgap here's another command that we need to run in the initial setup sidecar or script.

File: src/core/EdFi.DataManagementService.Core.Tests.Unit/Middleware/CoerceStringTypeMiddlewareTests.cs [hunk 10]
```diff
@@ -19,7 +19,7 @@
 
 namespace EdFi.DataManagementService.Core.Tests.Unit.Middleware
 {
-    public class CoerceStringTypeMiddlewareTests
+    public class CoerceFromStringsMiddlewareTests
```
Comment (simpat-adam) [inline src/core/EdFi.DataManagementService.Core.Tests.Unit/Middleware/CoerceStringTypeMiddlewareTests.cs:22]: Nit: you renamed the class but not the file.

File: src/core/EdFi.DataManagementService.Core/Handler/GetByIdHandler.cs [hunk 11]
```diff
@@ -3,14 +3,15 @@
 // The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 // See the LICENSE and NOTICES files in the project root for more information.
 
-using Microsoft.Extensions.Logging;
-using EdFi.DataManagementService.Core.Pipeline;
-using static EdFi.DataManagementService.Core.External.Backend.GetResult;
-using EdFi.DataManagementService.Core.External.Interface;
+using Polly;
 using EdFi.DataManagementService.Core.Backend;
 using EdFi.DataManagementService.Core.External.Backend;
```
Comment (simpat-adam) [inline src/core/EdFi.DataManagementService.Core/Handler/GetByIdHandler.cs:8]: This is unused

File: src/core/EdFi.DataManagementService.Core/Handler/UpsertHandler.cs [hunk 12]
```diff
@@ -14,6 +14,7 @@
 using Polly;
 using static EdFi.DataManagementService.Core.External.Backend.UpsertResult;
 using static EdFi.DataManagementService.Core.Response.FailureResponse;
+using static EdFi.DataManagementService.Core.Handler.Utility;
```
Comment (simpat-adam) [inline src/core/EdFi.DataManagementService.Core/Handler/UpsertHandler.cs:17]: `using EdFi.DataManagementService.Core.External.Backend;` is an unused using in this file

File: src/backend/EdFi.DataManagementService.Backend.OpenSearch/OpenSearchQueryHandlerRepository.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+using EdFi.DataManagementService.Core.External.Backend;
+using EdFi.DataManagementService.Core.External.Interface;
+using Microsoft.Extensions.Logging;
+using OpenSearch.Client;
+
+namespace EdFi.DataManagementService.Backend.OpenSearch;
+
+public class OpenSearchQueryHandlerRepository(
+    IOpenSearchClient openSearchClient,
+    ILogger<OpenSearchQueryHandlerRepository> _logger
```
Comment (simpat-adam) [inline src/backend/EdFi.DataManagementService.Backend.OpenSearch/OpenSearchQueryHandlerRepository.cs:15]: should be `logger` right?

**Review by simpat-adam — CHANGES_REQUESTED** [review]

**Review by simpat-adam — APPROVED** [review]

**Pull request merged**